File: _733_FloodFill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        check = image[sr][sc]
        if check == newColor:# It's possible they're the same, then will be endless loop
            return image
        
        def dfs(row, col):
            image[row][col] = newColor
            for row_dif, col_dif in ((-1, 0), (1, 0), (0, 1), (0, -1)):
                new_row, new_col = row + row_dif, col + col_dif
                if 0 <= new_row < len(image) and 0 <= new_col < len(image[0]) and image[new_row][new_col] == check:
                    dfs(new_row, new_col)
        
        dfs(sr, sc)
        return image

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        visited = set()
        max_row, max_col, color = len(image), len(image[0]), image[sr][sc]
        q = collections.deque()
        q.append((sr, sc))
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        while q:
            row, col = q.popleft()
            if 0 <= row < max_row and 0 <= col < max_col and image[row][col] == color and (row, col) not in visited:
                visited.add((row, col))# Don't visit back
                image[row][col] = newColor
                neighbors = [(row + direc[0], col + direc[1]) for direc in directions]
                q.extend(neighbors)
        return image

File: _733_FloodFill.java
class Solution {
    public void dfs(int[][] image, int r, int c, int color, int newColor) {
        // In case went into endless loop among same color neighbors.
        if (image[r][c] != color) {
            return;
        }
        image[r][c] = newColor;
        if (r >= 1) {
            dfs(image, r-1, c, color, newColor);
        }
        if (c >= 1) {
            dfs(image, r, c-1, color, newColor);
        }
        if (r+1 < image.length) {
            dfs(image, r+1, c, color, newColor);
        }
        if (c+1 < image[0].length) {
            dfs(image, r, c+1, color, newColor);
        }
    }

    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        // If already flood fill (color == image[sr][sc]), just return image
        if (color != image[sr][sc]) {
            dfs(image, sr, sc, image[sr][sc], color);;
        }
        return image;
    }
}